amaru/src/mempool/mod.rs

```diff
@@ -0,0 +1,104 @@
+use crate::ledger;
+use pallas_primitives::conway::Tx;
+use std::collections::BTreeSet;
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+// State
+// ----------------------------------------------------------------------------
+
+/// The state of the mempool is split into two sub-components:
+///
+/// - A _ledger_state_ which contains a view of a ledger state, which is used to validate
+///   transactions.
+///
+/// - An _ephemeral_ state, which is maintained as a sequence of diff operations to be applied on
+///   top of the _ledger_state_. It should be re-calculated as transactions are added or removed
+///   from the mempool.
+pub struct State<S, E>
+where
+    S: ledger::store::Store<Error = E>,
+{
+    /// The current view of the ledger state.
+    ledger_state: ledger::state::State<S, E>,
+
+    /// The ephemeral state, which is applied on top of the ledger_state using transactions
+    /// in the mempool.
+    ephemeral: ledger::state::VolatileDB,
+}
+
+/// A transaction in the mempool.
+pub trait MempoolTx<M>: Ord + PartialEq {
+    /// Any additional _metadata_ associated with the transaction. For example, it could include a timestamp
+    /// of when it was received, a reference to the peer that sent it, or some priority value used
+    /// to order transactions.
+    fn metadata(&self) -> Option<M> {
+        None
+    }
+
+    /// The transaction itself.
+    fn tx(&self) -> &Tx;
+}
+
+/// The mempool is a queue of transactions that are waiting to be validated and applied to an
+/// ephemeral ledger state.
+pub struct Mempool<T, M>
+where
+    T: MempoolTx<M>,
+{
+    /// The queue of transactions. The ordering of transactions are defined by the Ord trait of the
+    /// type T and left to the implementer.
+    transactions: Arc<RwLock<BTreeSet<T>>>,
```

Comment:
`Arc/RwLock` does not need to be documented as this is one of the most common patterns seen in async rust code for thread safety. People typically use either `Arc/Mutex` or `Arc/RwLock`. The former causes exactly one reader or one writer to access the data at a time. The later allows for a single writer, but multiple concurrent readers. Since we will have different mempools we are maintaining at once, each of them having simultaneous read access to the transactions makes the most sense.

I'll create a type alias for the BTreeSet as that will add clarity to the code.

Reply:
I wasn't clear in my comment (😂 ): I meant not to document the implementation, but the behaviour the implementation provides, eg. that `transactions` is thread-safe.